source/tutorials/learning-journey/packaging-existing-software.md

```diff
@@ -0,0 +1,531 @@
+
+To begin, you will download the [latest version](https://ftp.gnu.org/gnu/hello/hello-2.12.1.tar.gz) of `hello` using `fetchzip`, which takes the URI path to the download file and a SHA256 hash of its contents.
+
+:::{note}
+`fetchzip` can fetch [more archives](https://nixos.org/manual/nixpkgs/stable/#fetchurl) than just zip files!
+:::
+
+The hash cannot be known until after the tarball has been downloaded and unpacked, but Nix will complain if the hash supplied to `fetchzip` was incorrect, so it is common practice to supply a fake one with `lib.fakeSha256` and change the derivation definition after Nix reports the correct hash:
+
+```nix
+# hello.nix
+{ lib
+, stdenv
+, fetchzip
+}:
+
+stdenv.mkDerivation {
+  src = fetchzip {
+    url = "https://ftp.gnu.org/gnu/hello/hello-2.12.1.tar.gz";
+    sha256 = lib.fakeSha256;
+  };
+}
+```
+
+Save this file to `hello.nix` and try to build it with `nix-build`, observing your first build failure:
+
+```console
+$ nix-build hello.nix
+error: cannot evaluate a function that has an argument without a value ('lib')
+       Nix attempted to evaluate a function as a top level expression; in
+       this case it must have its arguments supplied either by default
+       values, or passed explicitly with '--arg' or '--argstr'. See
+       https://nixos.org/manual/nix/stable/language/constructs.html#functions.
+
+       at /home/nix-user/hello.nix:2:3:
+
+            1| # hello.nix
+            2| { lib
+             |   ^
+            3| , stdenv
+```
+
+Problem: the expression in `hello.nix` is a *function*, which only produces its intended output if it is passed the correct *arguments*.
+
+### A New Command
+`lib` is available from `nixpkgs`, which must be imported with another Nix expression in order to pass it as an argument to this derivation.
+
+The recommended way to do this is to create a `default.nix` in the same directory as `hello.nix`, with the following contents:
+
+```nix
+# default.nix
+let
+  pkgs = import <nixpkgs> { };
+in
+{
+  hello = pkgs.callPackage ./hello.nix { };
+}
+```
+
+This allows you to use `nix-build -A hello` to realize the derivation in `hello.nix`, similar to the current convention used in `nixpkgs`.
+
+:::{note}
+`callPackage` automatically passes attributes from `nixpkgs` to the given function, if they match attributes required by that function's argument attrset.
+
+In this case, `callPackage` will supply `lib`, and `stdenv` to the function defined in `hello.nix`.
+:::
+
+Now run the `nix-build` command with the new argument:
+
+```console
+$ nix-build -A hello
+error: derivation name missing
+```
+
+This new failure occurs with the *derivation*, further down in the file than the initial error on line 2 about the `lib` argument not having a value; the previous error was successfully resolved by changing the expression passed to `nix-build`.
```

Comment:
You can replace this entire sentence with something along the lines of "this is a _new_ error, let's explore where it comes from".